lvsfunc/source.py

```diff
@@ -0,0 +1,206 @@
+
+import os
+from abc import ABC, abstractmethod
+from io import BufferedReader
+from itertools import accumulate
+from pathlib import Path
+from subprocess import run
+from typing import Any, Callable, List, Optional, Tuple, Union, cast
+
+import vapoursynth as vs
+from pyparsedvd import vts_ifo
+from vsutil import is_image
+
+from .misc import get_matrix
+
+core = vs.core
+
+# List of containers that are better off being indexed externally
+
+
+annoying_formats = ['.iso', '.ts', '.vob']
+
+
+def source(file: str, ref: Optional[vs.VideoNode] = None,
+           force_lsmas: bool = False,
+           mpls: bool = False, mpls_playlist: int = 0, mpls_angle: int = 0,
+           **index_args: Any) -> vs.VideoNode:
+    """
+    Generic clip import function.
+    Automatically determines if ffms2 or L-SMASH should be used to import a clip, but L-SMASH can be forced.
+    It also automatically determines if an image has been imported.
+    You can set its fps using 'fpsnum' and 'fpsden', or using a reference clip with 'ref'.
+
+    Alias for this function is `lvsfunc.src`.
+
+    Dependencies:
+    * ffms2
+    * L-SMASH-Works (optional: m2ts sources or when forcing lsmas)
+    * d2vsource (optional: d2v sources)
+    * dgdecodenv (optional: dgi sources)
+    * VapourSynth-ReadMpls (optional: mpls sources)
+
+    :param file:              Input file
+    :param ref:               Use another clip as reference for the clip's format,
+                              resolution, and framerate (Default: None)
+    :param force_lsmas:       Force files to be imported with L-SMASH (Default: False)
+    :param mpls:              Load in a mpls file (Default: False)
+    :param mpls_playlist:     Playlist number, which is the number in mpls file name (Default: 0)
+    :param mpls_angle:        Angle number to select in the mpls playlist (Default: 0)
+    :param kwargs:            Arguments passed to the indexing filter
+
+    :return:                  Vapoursynth clip representing input file
+    """
+
+    # TODO: find a way to NOT have to rely on a million elif's
+    if file.startswith('file:///'):
+        file = file[8::]
+
+    # Error handling for some file types
+    if file.endswith('.mpls') and mpls is False:
+        raise ValueError("source: 'Set \"mpls = True\" and pass a path to the base Blu-ray directory for this kind of file'")  # noqa: E501
+    if os.path.splitext(file)[1].lower() in annoying_formats:
+        raise ValueError("source: 'Use an external indexer like d2vwitch or DGIndexNV for this kind of file'")  # noqa: E501
```

Comment:
Maybe have it switch to `dvd_source` here, or at least tell the user to use that instead?

Reply:
You should probably open a new PR or just a separate commit for this.